File: server/mpesa.py
import requests
import base64
import json
from datetime import datetime
import time
from db_setup import get_db_connection, dict_from_row
from mysql.connector import Error

# M-Pesa API credentials
CONSUMER_KEY = "sMwMwGZ8oOiSkNrUIrPbcCeWIO8UiQ3SV4CyX739uAyZVs1F"
CONSUMER_SECRET = "A3Hs5zRY3nDCn7XpxPuc1iAKpfy6UDdetiCalIAfuAIpgTROI5yCqqOewDfThh2o"
BUSINESS_SHORT_CODE = "174379"  # Lipa Na M-Pesa Shortcode
PASSKEY = "bfb279f9aa9bdbcf158e97dd71a467cd2e0c893059b10f78e6b72ada1ed2c919"
CALLBACK_URL = "https://webhook.site/3c1f62b5-4214-47d6-9f26-71c1f4b9c8f0"
API_BASE_URL = "https://sandbox.safaricom.co.ke"

def get_access_token():
    """Get OAuth access token from M-Pesa"""
    url = f"{API_BASE_URL}/oauth/v1/generate?grant_type=client_credentials"
    auth = base64.b64encode(f"{CONSUMER_KEY}:{CONSUMER_SECRET}".encode()).decode('utf-8')
    headers = {
        "Authorization": f"Basic {auth}"
    }
    
    try:
        response = requests.get(url, headers=headers)
        response_data = response.json()
        
        if "access_token" in response_data:
            return response_data["access_token"]
        else:
            print("Error getting access token:", response_data)
            return None
    except Exception as e:
        print(f"Exception while getting access token: {e}")
        return None

def generate_password():
    """Generate password for M-Pesa STK Push"""
    timestamp = datetime.now().strftime('%Y%m%d%H%M%S')
    password_str = f"{BUSINESS_SHORT_CODE}{PASSKEY}{timestamp}"
    password = base64.b64encode(password_str.encode()).decode('utf-8')
    return password, timestamp

def initiate_stk_push(phone_number, amount, account_reference, order_type, order_id, user_id):
    """Initiate STK Push to customer's phone"""
    access_token = get_access_token()
    if not access_token:
        return {"error": "Failed to get access token"}
    
    password, timestamp = generate_password()
    
    # Format phone number to match M-Pesa requirements
    if phone_number.startswith('+'):
        phone_number = phone_number[1:]
    if phone_number.startswith('0'):
        phone_number = '254' + phone_number[1:]
    
    url = f"{API_BASE_URL}/mpesa/stkpush/v1/processrequest"
    headers = {
        "Authorization": f"Bearer {access_token}",
        "Content-Type": "application/json"
    }
    
    payload = {
        "BusinessShortCode": BUSINESS_SHORT_CODE,
        "Password": password,
        "Timestamp": timestamp,
        "TransactionType": "CustomerPayBillOnline",
        "Amount": int(amount),
        "PartyA": int(phone_number),
        "PartyB": BUSINESS_SHORT_CODE,
        "PhoneNumber": int(phone_number),
        "CallBackURL": CALLBACK_URL,
        "AccountReference": account_reference,
        "TransactionDesc": f"Payment for {order_type} #{order_id}"
    }
    
    try:
        response = requests.post(url, json=payload, headers=headers)
        result = response.json()
        print(f"STK Push result: {result}")
        
        if "ResponseCode" in result and result["ResponseCode"] == "0":
            # Save transaction to database
            save_transaction_request(
                result["CheckoutRequestID"],
                result["MerchantRequestID"],
                order_type,
                order_id,
                user_id,
                amount,
                phone_number
            )
            
            return {
                "success": True,
                "checkoutRequestId": result["CheckoutRequestID"],
                "merchantRequestId": result["MerchantRequestID"]
            }
        else:
            return {
                "error": result.get("errorMessage", "STK Push failed"),
                "details": result
            }
    except Exception as e:
        print(f"Exception during STK Push: {e}")
        return {"error": str(e)}

def check_transaction_status(checkout_request_id):
    """Check status of an STK Push transaction"""
    connection = get_db_connection()
    if not connection:
        return {"error": "Database connection failed"}
    
    cursor = connection.cursor()
    
    try:
        # Check if transaction exists in database
        query = """
        SELECT * FROM mpesa_transactions 
        WHERE checkout_request_id = %s
        """
        cursor.execute(query, (checkout_request_id,))
        row = cursor.fetchone()
        
        if not row:
            return {"error": "Transaction not found"}
        
        transaction = dict_from_row(row, cursor)
        
        # If transaction is still pending, check status from M-Pesa
        if transaction["status"] == "pending":
            access_token = get_access_token()
            if not access_token:
                return {"error": "Failed to get access token"}
            
            password, timestamp = generate_password()
            
            url = f"{API_BASE_URL}/mpesa/stkpushquery/v1/query"
            headers = {
                "Authorization": f"Bearer {access_token}",
                "Content-Type": "application/json"
            }
            
            payload = {
                "BusinessShortCode": BUSINESS_SHORT_CODE,
                "Password": password,
                "Timestamp": timestamp,
                "CheckoutRequestID": checkout_request_id
            }
            
            try:
                response = requests.post(url, json=payload, headers=headers)
                result = response.json()
                print(f"Transaction status query result: {result}")
                
                if "ResultCode" in result:
                    if result["ResultCode"] == "0":
                        # Update transaction status to completed
                        update_transaction_status(
                            checkout_request_id,
                            "completed",
                            result.get("ResultCode"),
                            result.get("ResultDesc")
                        )
                        
                        # Update order status
                        update_order_status(
                            transaction["order_type"],
                            transaction["order_id"],
                            "completed"
                        )
                        
                        return {
                            "success": True,
                            "status": "completed",
                            "message": "Payment completed successfully"
                        }
                    else:
                        # Update transaction status to failed
                        update_transaction_status(
                            checkout_request_id,
                            "failed",
                            result.get("ResultCode"),
                            result.get("ResultDesc")
                        )
                        
                        return {
                            "success": False,
                            "status": "failed",
                            "message": result.get("ResultDesc", "Payment failed")
                        }
                else:
                    return {
                        "status": "pending",
                        "message": "Payment is being processed"
                    }
            except Exception as e:
                print(f"Exception during status check: {e}")
                return {"error": str(e)}
        else:
            # Return status from database
            return {
                "status": transaction["status"],
                "message": transaction["result_desc"] if transaction["result_desc"] else 
                          "Payment completed" if transaction["status"] == "completed" else "Payment failed"
            }
    except Exception as e:
        print(f"Error checking transaction: {e}")
        return {"error": str(e)}
    finally:
        if connection.is_connected():
            cursor.close()
            connection.close()

def save_transaction_request(checkout_request_id, merchant_request_id, order_type, order_id, user_id, amount, phone_number):
    """Save M-Pesa transaction request to database"""
    connection = get_db_connection()
    if not connection:
        return False
    
    cursor = connection.cursor()
    
    try:
        query = """
        INSERT INTO mpesa_transactions
        (checkout_request_id, merchant_request_id, order_type, order_id, user_id, amount, phone_number)
        VALUES (%s, %s, %s, %s, %s, %s, %s)
        """
        cursor.execute(query, (
            checkout_request_id,
            merchant_request_id,
            order_type,
            order_id,
            user_id,
            amount,
            phone_number
        ))
        connection.commit()
        return True
    except Error as e:
        print(f"Error saving transaction: {e}")
        return False
    finally:
        if connection.is_connected():
            cursor.close()
            connection.close()

def update_transaction_status(checkout_request_id, status, result_code=None, result_desc=None):
    """Update M-Pesa transaction status in database"""
    connection = get_db_connection()
    if not connection:
        return False
    
    cursor = connection.cursor()
    
    try:
        query = """
        UPDATE mpesa_transactions
        SET status = %s, result_code = %s, result_desc = %s
        WHERE checkout_request_id = %s
        """
        cursor.execute(query, (status, result_code, result_desc, checkout_request_id))
        connection.commit()
        return True
    except Error as e:
        print(f"Error updating transaction: {e}")
        return False
    finally:
        if connection.is_connected():
            cursor.close()
            connection.close()

def update_order_status(order_type, order_id, payment_status):
    """Update order payment status in database"""
    connection = get_db_connection()
    if not connection:
        return False
    
    cursor = connection.cursor()
    
    try:
        if order_type == "artwork":
            query = """
            UPDATE artwork_orders
            SET payment_status = %s
            WHERE id = %s
            """
        elif order_type == "exhibition":
            query = """
            UPDATE exhibition_bookings
            SET payment_status = %s
            WHERE id = %s
            """
        else:
            return False
        
        cursor.execute(query, (payment_status, order_id))
        connection.commit()
        
        # If it's an artwork order and payment is completed, update artwork status
        if order_type == "artwork" and payment_status == "completed":
            query = """
            UPDATE artworks a
            JOIN artwork_orders o ON a.id = o.artwork_id
            SET a.status = 'sold'
            WHERE o.id = %s
            """
            cursor.execute(query, (order_id,))
            connection.commit()
        
        # If it's an exhibition booking and payment is completed, update available slots
        if order_type == "exhibition" and payment_status == "completed":
            query = """
            UPDATE exhibitions e
            JOIN exhibition_bookings b ON e.id = b.exhibition_id
            SET e.available_slots = e.available_slots - b.slots
            WHERE b.id = %s
            """
            cursor.execute(query, (order_id,))
            connection.commit()
        
        return True
    except Error as e:
        print(f"Error updating order: {e}")
        return False
    finally:
        if connection.is_connected():
            cursor.close()
            connection.close()

def handle_mpesa_callback(callback_data):
    """Handle M-Pesa callback data"""
    try:
        checkout_request_id = callback_data.get("CheckoutRequestID")
        result_code = callback_data.get("ResultCode")
        result_desc = callback_data.get("ResultDesc")
        
        if not checkout_request_id:
            return {"error": "Missing CheckoutRequestID"}
        
        connection = get_db_connection()
        if not connection:
            return {"error": "Database connection failed"}
        
        cursor = connection.cursor()
        
        try:
            # Get order details from orders table
            query = """
            SELECT id, user_id, order_type, item_id, amount 
            FROM orders 
            WHERE checkout_request_id = %s
            """
            cursor.execute(query, (checkout_request_id,))
            order = cursor.fetchone()
            
            if not order:
                return {"error": "Order not found"}
            
            order_id, user_id, order_type, item_id, amount = order
            
            # Update order status based on M-Pesa response
            status = "completed" if result_code == "0" else "failed"
            
            update_query = """
            UPDATE orders 
            SET status = %s
            WHERE checkout_request_id = %s
            """
            cursor.execute(update_query, (status, checkout_request_id))
            
            # If payment successful and it's an exhibition booking, create ticket
            if status == "completed" and order_type == "exhibition":
                # Get slots from exhibition_bookings
                cursor.execute("""
                    SELECT slots FROM exhibition_bookings 
                    WHERE id = %s
                """, (item_id,))
                booking = cursor.fetchone()
                
                if booking:
                    slots = booking[0]
                    create_ticket(order_id, user_id, item_id, slots)
                    
                    # Update exhibition available slots
                    cursor.execute("""
                        UPDATE exhibitions 
                        SET available_slots = available_slots - %s 
                        WHERE id = %s
                    """, (slots, item_id))
            
            connection.commit()
            return {"success": True, "status": status}
            
        finally:
            cursor.close()
            connection.close()
            
    except Exception as e:
        print(f"Error handling M-Pesa callback: {e}")
        return {"error": str(e)}

def handle_stk_push_request(request_data):
    """Handle STK Push request from frontend"""
    try:
        phone_number = request_data.get("phoneNumber")
        amount = request_data.get("amount")
        order_type = request_data.get("orderType")
        item_id = request_data.get("orderId")
        user_id = request_data.get("userId")
        account_reference = request_data.get("accountReference")
        
        if not all([phone_number, amount, order_type, item_id, user_id, account_reference]):
            return {"error": "Missing required fields"}
        
        # Initiate STK Push
        stk_response = initiate_stk_push(
            phone_number, amount, account_reference, 
            order_type, item_id, user_id
        )
        
        if "error" in stk_response:
            return stk_response
            
        # Create order in database
        create_order(
            user_id=user_id,
            order_type=order_type,
            item_id=item_id,
            amount=amount,
            checkout_request_id=stk_response.get("checkoutRequestId"),
            merchant_request_id=stk_response.get("merchantRequestId")
        )
        
        return stk_response
        
    except Exception as e:
        print(f"Error handling STK Push request: {e}")
        return {"error": str(e)}

File: src/utils/mpesa.ts
// M-Pesa API utilities

// M-Pesa credentials
const CONSUMER_KEY = "sMwMwGZ8oOiSkNrUIrPbcCeWIO8UiQ3SV4CyX739uAyZVs1F";
const CONSUMER_SECRET = "A3Hs5zRY3nDCn7XpxPuc1iAKpfy6UDdetiCalIAfuAIpgTROI5yCqqOewDfThh2o";
const API_URL = 'http://localhost:8000';
const CALLBACK_URL = "https://webhook.site/3c1f62b5-4214-47d6-9f26-71c1f4b9c8f0";
const API_BASE_URL = "https://sandbox.safaricom.co.ke";

// Function to initiate STK Push
export const initiateSTKPush = async (
  phoneNumber: string,
  amount: number,
  orderType: 'artwork' | 'exhibition',
  orderId: string,
  accountReference: string
): Promise<any> => {
  try {
    // Format phone number to match M-Pesa requirements (remove '+' if present)
    const formattedPhone = phoneNumber.startsWith('+') 
      ? phoneNumber.substring(1) 
      : phoneNumber;
    
    // Add validation for Kenyan phone numbers
    // Should start with 254 (country code)
    let validatedPhone = formattedPhone;
    if (validatedPhone.startsWith('0')) {
      // Convert format from 07XXXXXXXX to 2547XXXXXXXX
      validatedPhone = '254' + validatedPhone.substring(1);
    } else if (!validatedPhone.startsWith('254')) {
      // Add country code if missing
      validatedPhone = '254' + validatedPhone;
    }
    
    console.log(`Initiating STK Push for phone: ${validatedPhone}, amount: ${amount}, order type: ${orderType}`);
    
    // Get user ID from localStorage
    const userId = localStorage.getItem('userId') || '';
    
    // If userId is missing, throw an error
    if (!userId) {
      throw new Error("User ID is missing. Please log in and try again.");
    }
    
    const requestBody = {
      phoneNumber: validatedPhone,
      amount,
      orderType,
      orderId,
      userId,
      accountReference,
      callbackUrl: CALLBACK_URL
    };
    
    console.log('STK Push request body:', requestBody);
    
    const response = await fetch(`${API_URL}/mpesa/stk-push`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify(requestBody),
    });
    
    if (!response.ok) {
      const errorData = await response.json();
      console.error('STK Push failed with server response:', errorData);
      throw new Error(`M-Pesa API Error: ${errorData.error || 'Unknown error'}`);
    }
    
    const responseData = await response.json();
    console.log('STK Push response:', responseData);
    return responseData;
  } catch (error) {
    console.error('M-Pesa STK Push error:', error);
    throw error;
  }
};

// Function to check transaction status
export const checkTransactionStatus = async (checkoutRequestId: string): Promise<any> => {
  try {
    console.log(`Checking transaction status for checkoutRequestId: ${checkoutRequestId}`);
    
    const response = await fetch(`${API_URL}/mpesa/status/${checkoutRequestId}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });
    
    if (!response.ok) {
      const errorData = await response.json();
      console.error('Transaction status check failed with server response:', errorData);
      throw new Error(`M-Pesa API Error: ${errorData.error || 'Unknown error'}`);
    }
    
    const responseData = await response.json();
    console.log('Transaction status response:', responseData);
    return responseData;
  } catch (error) {
    console.error('M-Pesa transaction status check error:', error);
    throw error;
  }
};

// Update finalizeOrder to handle both artwork orders and exhibition tickets
export const finalizeOrder = async (
  checkoutRequestId: string,
  orderType: 'artwork' | 'exhibition',
  orderData: any
): Promise<any> => {
  try {
    const endpoint = orderType === 'artwork' ? '/orders/finalize' : '/tickets/finalize';
    const response = await fetch(`${API_URL}${endpoint}`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
      },
      body: JSON.stringify({
        checkoutRequestId,
        orderType,
        orderData
      }),
    });
    
    return await response.json();
  } catch (error) {
    console.error('Order finalization error:', error);
    throw error;
  }
};

// Update getUserOrders to fetch both orders and tickets
export const getUserOrders = async (userId: string): Promise<any> => {
  try {
    const [ordersResponse, ticketsResponse] = await Promise.all([
      fetch(`${API_URL}/orders/user/${userId}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      }),
      fetch(`${API_URL}/tickets/user/${userId}`, {
        method: 'GET',
        headers: {
          'Content-Type': 'application/json',
        },
      })
    ]);
    
    const [orders, tickets] = await Promise.all([
      ordersResponse.json(),
      ticketsResponse.json()
    ]);
    
    return {
      orders: orders.orders || [],
      tickets: tickets.tickets || []
    };
  } catch (error) {
    console.error('Get user orders/tickets error:', error);
    throw error;
  }
};

// Function to generate exhibition ticket
export const generateExhibitionTicket = async (bookingId: string): Promise<any> => {
  try {
    const response = await fetch(`${API_URL}/tickets/generate/${bookingId}`, {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
      },
    });
    
    return await response.json();
  } catch (error) {
    console.error('Ticket generation error:', error);
    throw error;
  }
};
